### PR TITLE
hexpatch: init at 1.9.3

### LINKED ISSUE #346039: Package request: hexpatch
**HexPatch: a binary patcher and editor written in Rust with terminal user interface (TUI)**

HexPatch is a binary patcher and editor with terminal user interface, capable of disassembling instructions and assembling patches. It supports a variety of architectures and file formats, and can edit remote files via SSH.

**Metadata**

* homepage URL: https://etto48.github.io/HexPatch/
* source URL: https://github.com/Etto48/HexPatch
* license: mit
* platforms: unix, linux, darwin

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

### PR DESCRIPTION
**Pull request by vandrw**
Added [hexpatch](https://github.com/Etto48/HexPatch), a binary patcher and editor written in Rust with terminal user interface.

Closes #346039.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/tools/misc/hexpatch/default.nix [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+  llvmPackages,
+}:
+
+rustPlatform.buildRustPackage rec {
+  pname = "hexpatch";
+  version = "1.9.0";
+
+  src = fetchFromGitHub {
+    owner = "Etto48";
+    repo = pname;
```
Comment (AndersonTorres) [inline pkgs/tools/misc/hexpatch/default.nix:18]:
```suggestion
    repo = "hexpatch";
```

File: pkgs/tools/misc/hexpatch/default.nix [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+  };
+
+  cargoHash = "sha256-Bp6+y3ZmkAGxkcz4GJk9tBtglltNPhnGk4lLV9SgDks=";
+
+  nativeBuildInputs =
+    [
+      cmake
+      python3
+    ]
+    ++ lib.optionals (stdenv.system == "x86_64-linux" || stdenv.system == "aarch64-linux") [ gcc ] # Linux
```
Comment (AndersonTorres) [inline pkgs/tools/misc/hexpatch/default.nix:30]:
```suggestion
    ++ lib.optionals (stdenv.system == "x86_64-linux" || stdenv.system == "aarch64-linux") [ gcc ]
```

File: pkgs/tools/misc/hexpatch/default.nix [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+    ++ lib.optionals (
+      stdenv.system == "x86_64-darwin"
+      || stdenv.system == "aarch64-darwin"
+      || stdenv.system == "x86_64-windows"
+      || stdenv.system == "aarch64-windows"
+    ) [ llvmPackages.libclang ]; # MacOS and Windows
```
Comment (AndersonTorres) [inline pkgs/tools/misc/hexpatch/default.nix:36]:
```suggestion
    ) [ llvmPackages.libclang ];
```

File: pkgs/tools/misc/hexpatch/default.nix [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+
+  # Rename the binary from hex-patch to hexpatch
```
Comment (AndersonTorres) [inline pkgs/tools/misc/hexpatch/default.nix:38]:
```suggestion
```

File: pkgs/tools/misc/hexpatch/default.nix [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+  postFixup = ''
+    mv $out/bin/hex-patch $out/bin/hexpatch
```
Comment (AndersonTorres) [inline pkgs/tools/misc/hexpatch/default.nix:40]: Use a symlink instead.

File: pkgs/tools/misc/hexpatch/default.nix [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+  '';
+
+  meta = with lib; {
```
Comment (AndersonTorres) [inline pkgs/tools/misc/hexpatch/default.nix:43]:
```suggestion
  meta = {
```

File: pkgs/tools/misc/hexpatch/default.nix [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+    description = "A binary patcher and editor written in Rust with terminal user interface.";
+    longDescription = ''
+      HexPatch is a binary patcher and editor with terminal user interface (TUI),
+      capable of disassembling instructions and assembling patches. It supports a
+      variety of architectures and file formats, and can edit remote files
+      via SSH.
+    '';
+    homepage = "https://etto48.github.io/HexPatch/";
+    license = licenses.mit;
+    maintainers = with maintainers; [ Etto48 ];
```
Comment (AndersonTorres) [inline pkgs/tools/misc/hexpatch/default.nix:53]:
```suggestion
    license = lib.licenses.mit;
    maintainers = with lib.maintainers; [ Etto48 ];
```

File: pkgs/tools/misc/hexpatch/default.nix [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+  cargoHash = "sha256-Bp6+y3ZmkAGxkcz4GJk9tBtglltNPhnGk4lLV9SgDks=";
+
+  nativeBuildInputs =
+    [
+      cmake
+      python3
+    ]
+    ++ lib.optionals (stdenv.system == "x86_64-linux" || stdenv.system == "aarch64-linux") [ gcc ] # Linux
+    ++ lib.optionals (
+      stdenv.system == "x86_64-darwin"
```
Comment (AndersonTorres) [inline pkgs/tools/misc/hexpatch/default.nix:32]: Instead of this, you should use `stdenv.hostPlatform.isWindows`  etc.

**Review by AndersonTorres — CHANGES_REQUESTED** [review]

**Comment by vandrw** [comment]
Hi @AndersonTorres. Thanks a lot for reviewing this! I've implemented your suggestions, and I made a new commit. Please let me know if anything else needs adjustment now.

Regarding hexpatch, it requires [MSVC](https://visualstudio.microsoft.com/visual-cpp-build-tools/) on Windows, and [Xcode Command Line Tools](https://developer.apple.com/) on macOS. I used LLVM in this case, but is there another way to call these specifically? I think it could potentially improve build time on these platforms, but I was not sure.

File: pkgs/tools/misc/hexpatch/default.nix
Comment (Scrumplex) [inline pkgs/tools/misc/hexpatch/default.nix:1]: Could you move this package to `pkgs/by-name/he/hexpatch/package.nix`?

You just added this .nix file without referencing it in `pkgs/top-level/all-packages.nix`. New packages should go into `pkgs/by-name` where they will be picked up automatically.

**Comment by vandrw** [comment]
Hi @Scrumplex ! Thanks for the suggestions. I have moved the package to the suggested location and updated it to the latest version. Please let me know if you see anything else that needs fixes.

File: pkgs/by-name/he/hexpatch/package.nix [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+    description = "A binary patcher and editor written in Rust with terminal user interface.";
+    longDescription = ''
+      HexPatch is a binary patcher and editor with terminal user interface (TUI),
+      capable of disassembling instructions and assembling patches. It supports a
+      variety of architectures and file formats, and can edit remote files
+      via SSH.
+    '';
+    homepage = "https://etto48.github.io/HexPatch/";
+    license = lib.licenses.mit;
+    maintainers = with lib.maintainers; [ Etto48 ];
```
Comment (Scrumplex) [inline pkgs/by-name/he/hexpatch/package.nix:49]: You need to create an entry in [`maintainers/maintainer-list.nix`](https://github.com/NixOS/nixpkgs/blob/master/maintainers/README.md) for this to work. Make sure to do it in a separate commit.

Comment (vandrw) [reply pkgs/by-name/he/hexpatch/package.nix]: I see! Apologies, I assumed that this would be the software maintainer, and not the Nix package one. I will add myself to the list, and let the main developer step in if he wishes to maintain it later.

Thank you very much for your guidance!

File: pkgs/by-name/he/hexpatch/package.nix [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+  version = "1.9.3";
+
+  src = fetchFromGitHub {
+    owner = "Etto48";
+    repo = "HexPatch";
+    rev = "v${version}";
+    hash = "sha256-yyYxRJ+o+Z5z7PmjcFCsahRXZ9JHFmGmituzGTxY6ec=";
+  };
+
+  cargoHash = "sha256-bckZtHzgKQow7l2Cg6Xi1oNk13FH2xKe8cbZ/Xj9E04=";
```
Comment (Scrumplex) [inline pkgs/by-name/he/hexpatch/package.nix:23]: These hashes need to be updated. I assume this built fine for you. That usually happens because Nix doesn't actually verify the hashes again if they are already cached locally. You can force a refetch by setting these hashes to an empty string.

Nix should then report the new (correct) hashes.

File: pkgs/by-name/he/hexpatch/package.nix [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+
+  nativeBuildInputs =
+    [
+      cmake
+      python3
+    ]
+    ++ lib.optionals stdenv.hostPlatform.isLinux [ gcc ]
+    ++ lib.optionals (stdenv.hostPlatform.isWindows || stdenv.hostPlatform.isDarwin) [
+      llvmPackages.libclang
+    ];
```
Comment (Scrumplex) [inline pkgs/by-name/he/hexpatch/package.nix:33]: These should not be needed, as they are already provided by Nixpkgs' `stdenv`. I tried building this package on `x86_64-linux` and `aarch64-darwin` without these and it worked just fine.

```suggestion
  nativeBuildInputs = [
    cmake
    python3
  ];
```

File: pkgs/by-name/he/hexpatch/package.nix [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+
+  postFixup = ''
+    ln -s $out/bin/hex-patch $out/bin/hexpatch
+  '';
+
+  meta = {
+    description = "A binary patcher and editor written in Rust with terminal user interface.";
```
Comment (Scrumplex) [inline pkgs/by-name/he/hexpatch/package.nix:40]: We try to avoid an indefinite article in descriptions.

See https://github.com/NixOS/nixpkgs/tree/00cd61f517aa325478e9f743d572f65d2bbcb16a/pkgs#meta-attributes and/or https://github.com/NixOS/nixpkgs/pull/317959

```suggestion
    description = "Binary patcher and editor written in Rust with terminal user interface.";
```

**Review by Scrumplex — COMMENTED** [review]
You should make sure to squash your commits.

I would suggest that this PR should have the following commits:
1. `maintainers: add vandrw as a maintainer`
2. `hexpatch: init at 1.9.3`

If you are not sure how to do that with Git, I can assist you there. The keyword here is git rebase.

**Comment by vandrw** [comment]
Apologies for the crazy labels and requests! It seems that the commit history was too borked for a rebase to work 😅
